site/content/features/virtual-threads.md

```diff
@@ -0,0 +1,433 @@
+
+## Rate Limiting
+
+Virtual threads improve application throughput since you can have many more concurrent tasks than with platform threads. That can put pressure on the services that the tasks invoke. For example, a web service may not tolerate huge numbers of concurrent requests.
+
+With platform threads, an easy (if crude) tuning factor is the size of the thread pool for those tasks. But you should not pool virtual threads. Scheduling tasks on virtual threads that are then scheduled on platform threads is clearly inefficient. And what is the upside? To limit the number virtual threads to the smallish number of concurrent requests that your service tolerates? Then why are you using virtual threads in the first place? 
+
+With virtual threads, you should use alternative mechanisms for controlling access to limited resources. Instead of an overall limit on concurrent tasks, protect each resource in an appropriate way. For database connections, the connection pool may already do the right thing. When accessing a web service, you know your service, and can provide appropriate rate limiting.
+
+As an example, on my personal web site, I provide demo services for producing random items. If a large number of requests comes at an instant from the same IP address, the hosting company blacklists the IP address. 
+
+The following sandbox shows rate limiting with a simple semaphore that allows a small number of concurrent requests. When the maximum is exceeded, the `acquire` method blocks, but that is ok. With virtual threads, blocking is cheap.
+
+{{< sandbox version=java21 preview="true" mainclass="RateLimitDemo" >}}{{< sandboxsource "RateLimitDemo.java" >}}
+import java.util.*;
+import java.util.concurrent.*;
+import java.net.*;
+import java.net.http.*;
+
+public class RateLimitDemo {
+   public static void main(String[] args) throws InterruptedException, ExecutionException {
+      ExecutorService service = Executors.newVirtualThreadPerTaskExecutor();
+      List<Future<String>> futures = new ArrayList<>();
+      final int TASKS = 250;
+      for (int i = 1; i <= TASKS; i++)
+         futures.add(service.submit(() -> get("https://horstmann.com/random/word")));
+      for (Future<String> f : futures)
+         System.out.print(f.get() + " ");
+      System.out.println();
+      service.close();
+   }
+
+   private static HttpClient client = HttpClient.newHttpClient();
+
+   private static final Semaphore SEMAPHORE = new Semaphore(20);
+   
+   public static String get(String url) {
+      try {
+         var request = HttpRequest.newBuilder().uri(new URI(url)).GET().build();
+         SEMAPHORE.acquire();
+         try {
+            Thread.sleep(100);
+            return client.send(request, HttpResponse.BodyHandlers.ofString()).body();
+         } finally {
+            SEMAPHORE.release();
+         }
+      } catch (Exception ex) {
+         ex.printStackTrace();
```

Comment:
Why printing the stack trace here? It should be printed anyways when `Future::get` is called.